### ISSUE TITLE
Rust Bindings for ModemManager

### ISSUE THREAD

**Issue by stfl** [gtk-rs/gir#637]
I am trying to generate bindings for libmm-glib [libmm-glib](https://www.freedesktop.org/software/ModemManager/libmm-glib/1.8.0/index.html). Various issues came up. Some of them I figured out but others I don't know how to progress.

I downloaded the source for modemmanager (1.8.10), generated the ModemManager.gir file. I am on Fedora 27 with gobject-introspection 1.54.1. [ModemManager.gir](https://paste.ee/p/MZKde)
I commented out some lines where rust-gir reported a wrong formatting. (at line 1924) I already reported this to the ModemManager maintainer.

I generated the rust FFI bindings with:

    cargo run --release -- -c Gir_mm_sys.toml -d ../gir-files -m sys -o ../mm-sys

*Gir_mm_sys.toml*:
```toml
[options]
work_mode = "sys"
girs_dir = "../gir-files"
library = "ModemManager"
version = "1.0"
min_cfg_version = "1.6.0"
# target_path = "."
# generate_safety_asserts = false
single_version_file = "."

external_libraries = [
   "GLib",
   "Gio",
   "GObject",
]

ignore = [
  "MMCall",
  "MM_SIMPLE_PROPERTY_STATE",
  "MM_SIMPLE_PROPERTY_SIGNAL_QUALITY",
  "MM_SIMPLE_PROPERTY_CURRENT_BANDS",
  "MM_SIMPLE_PROPERTY_ACCESS_TECHNOLOGIES",
  "MM_SIMPLE_PROPERTY_3GPP_REGISTRATION_STATE",
  "MM_SIMPLE_PROPERTY_3GPP_OPERATOR_CODE",
  "MM_SIMPLE_PROPERTY_3GPP_OPERATOR_NAME",
  "MM_SIMPLE_PROPERTY_3GPP_SUBSCRIPTION_STATE",
  "MM_SIMPLE_PROPERTY_CDMA_CDMA1X_REGISTRATION_STATE",
  "MM_SIMPLE_PROPERTY_CDMA_EVDO_REGISTRATION_STATE",
  "MM_SIMPLE_PROPERTY_CDMA_SID",
  "MM_SIMPLE_PROPERTY_CDMA_NID",
]

```

I ignored MMCall because that's where the syntax error in the .gir file appeared..

Requirements for libmm are according to fedora 27 package. 

```
glib2 >= 2.32
libc.so.6(GLIBC_2.14)(64bit)
libgio-2.0.so.0()(64bit)
libglib-2.0.so.0()(64bit)
libgobject-2.0.so.0()(64bit)
libpthread.so.0()(64bit)
libpthread.so.0(GLIBC_2.2.5)(64bit)
rtld(GNU_HASH)
```

So GLib, Gio and GObject should do for "external_libraries" right?
I had to update the required package versions to "0.6.0". The default generated was "0.2.0" which didn't work.

Now the binding compile but when running `cargo test` I get a couple of errors. Some of the constants cannot be matched because they are present in ModemManager.gir but not actually in the installed libmm-glib.h. I ignored these constants but it still does runs the test and shows a reference in src/lib.rs.

The same for MMCall but I assume that rust-gir cannot properly generate the struct because of the problem in ModemManager.gir
(I also tried ignoring ModemManager.MMCall with not changes)

```
   Compiling modem-manager-sys v0.2.0 (file:///home/sll/work/mee032-weinmann/rusteval/modem-manager-sys)
    Finished dev [unoptimized + debuginfo] target(s) in 1.68s
     Running /home/sll/work/mee032-weinmann/rusteval/modem-manager-sys/target/debug/deps/modem_manager_sys-4b7eccccce5c1f56

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running /home/sll/work/mee032-weinmann/rusteval/modem-manager-sys/target/debug/deps/abi-14909627a1f6da49

running 2 tests
test cross_validate_layout_with_c ... FAILED
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_3GPP_OPERATOR_CODE’ undeclared (first use in this function); did you mean ‘MM_SIM_PROPERTY_OPERATORNAME’?
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_3GPP_OPERATOR_NAME’ undeclared (first use in this function); did you mean ‘MM_SIM_PROPERTY_OPERATORNAME’?
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_3GPP_REGISTRATION_STATE’ undeclared (first use in this function); did you mean ‘MM_TYPE_MODEM_3GPP_REGISTRATION_STATE’?
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_3GPP_SUBSCRIPTION_STATE’ undeclared (first use in this function); did you mean ‘MM_TYPE_MODEM_3GPP_SUBSCRIPTION_STATE’?
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_ACCESS_TECHNOLOGIES’ undeclared (first use in this function); did you mean ‘MM_MODEM_PROPERTY_ACCESSTECHNOLOGIES’?
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_CDMA_CDMA1X_REGISTRATION_STATE’ undeclared (first use in this function); did you mean ‘MM_TYPE_MODEM_CDMA_REGISTRATION_STATE’?
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_CDMA_EVDO_REGISTRATION_STATE’ undeclared (first use in this function); did you mean ‘MM_TYPE_MODEM_CDMA_REGISTRATION_STATE’?
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_CDMA_NID’ undeclared (first use in this function); did you mean ‘MM_SIM_PROPERTY_IMSI’?
tests/constant.c:9:22: note: in expansion of macro ‘ABI_CONSTANT_NAME’
     printf(_Generic((ABI_CONSTANT_NAME),
                      ^~~~~~~~~~~~~~~~~
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
tests/constant.c:9:22: note: in expansion of macro ‘ABI_CONSTANT_NAME’
     printf(_Generic((ABI_CONSTANT_NAME),
                      ^~~~~~~~~~~~~~~~~
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_CDMA_SID’ undeclared (first use in this function); did you mean ‘MM_SIM_PROPERTY_IMSI’?
tests/constant.c:9:22: note: in expansion of macro ‘ABI_CONSTANT_NAME’
     printf(_Generic((ABI_CONSTANT_NAME),
                      ^~~~~~~~~~~~~~~~~
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
tests/constant.c:9:22: note: in expansion of macro ‘ABI_CONSTANT_NAME’
     printf(_Generic((ABI_CONSTANT_NAME),
                      ^~~~~~~~~~~~~~~~~
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_CURRENT_BANDS’ undeclared (first use in this function); did you mean ‘MM_MODEM_PROPERTY_CURRENTBANDS’?
tests/constant.c:9:22: note: in expansion of macro ‘ABI_CONSTANT_NAME’
     printf(_Generic((ABI_CONSTANT_NAME),
                      ^~~~~~~~~~~~~~~~~
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
tests/constant.c:9:22: note: in expansion of macro ‘ABI_CONSTANT_NAME’
     printf(_Generic((ABI_CONSTANT_NAME),
                      ^~~~~~~~~~~~~~~~~
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_SIGNAL_QUALITY’ undeclared (first use in this function); did you mean ‘MM_MODEM_PROPERTY_SIGNALQUALITY’?
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
<command-line>:0:19: note: in definition of macro ‘ABI_CONSTANT_NAME’
tests/constant.c: In function ‘main’:
<command-line>:0:19: error: ‘MM_SIMPLE_PROPERTY_STATE’ undeclared (first use in this function); did you mean ‘MM_SMS_PROPERTY_STATE’?
tests/constant.c:9:22: note: in expansion of macro ‘ABI_CONSTANT_NAME’
     printf(_Generic((ABI_CONSTANT_NAME),
                      ^~~~~~~~~~~~~~~~~
<command-line>:0:19: note: each undeclared identifier is reported only once for each function it appears in
tests/constant.c:9:22: note: in expansion of macro ‘ABI_CONSTANT_NAME’
     printf(_Generic((ABI_CONSTANT_NAME),
                      ^~~~~~~~~~~~~~~~~
test cross_validate_constants_with_c ... test cross_validate_constants_with_c has been running for over 60 seconds
test cross_validate_constants_with_c ... FAILED

failures:

---- cross_validate_layout_with_c stdout ----
Layout mismatch for MMCall
Rust: Layout { size: 8, alignment: 8 }
C:    Layout { size: 48, alignment: 8 }
layout    ... 24 passed; 1 failed (compilation errors: 0)
layout    ... 49 passed; 1 failed (compilation errors: 0)
layout    ... 74 passed; 1 failed (compilation errors: 0)
layout    ... 99 passed; 1 failed (compilation errors: 0)
layout    ... 124 passed; 1 failed (compilation errors: 0)
layout    ... 149 passed; 1 failed (compilation errors: 0)
layout    ... 174 passed; 1 failed (compilation errors: 0)
layout    ... 199 passed; 1 failed (compilation errors: 0)
thread 'cross_validate_layout_with_c' panicked at 'FAILED: 200 passed; 1 failed (compilation errors: 0)', tests/abi.rs:126:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- cross_validate_constants_with_c stdout ----
Constant value mismatch for MM_BEARER_IP_FAMILY_ANY
Rust: "-1"
C:    "4294967295"
constants ... 24 passed; 1 failed (compilation errors: 0)
constants ... 49 passed; 1 failed (compilation errors: 0)
constants ... 74 passed; 1 failed (compilation errors: 0)
constants ... 99 passed; 1 failed (compilation errors: 0)
constants ... 124 passed; 1 failed (compilation errors: 0)
constants ... 149 passed; 1 failed (compilation errors: 0)
constants ... 174 passed; 1 failed (compilation errors: 0)
constants ... 199 passed; 1 failed (compilation errors: 0)
Constant value mismatch for MM_MODEM_ACCESS_TECHNOLOGY_ANY
Rust: "-1"
C:    "4294967295"
constants ... 223 passed; 2 failed (compilation errors: 0)
constants ... 248 passed; 2 failed (compilation errors: 0)
constants ... 273 passed; 2 failed (compilation errors: 0)
constants ... 298 passed; 2 failed (compilation errors: 0)
Constant value mismatch for MM_MODEM_CAPABILITY_ANY
Rust: "-1"
C:    "4294967295"
constants ... 322 passed; 3 failed (compilation errors: 0)
constants ... 347 passed; 3 failed (compilation errors: 0)
constants ... 372 passed; 3 failed (compilation errors: 0)
constants ... 397 passed; 3 failed (compilation errors: 0)
Constant value mismatch for MM_MODEM_MODE_ANY
Rust: "-1"
C:    "4294967295"
constants ... 421 passed; 4 failed (compilation errors: 0)
constants ... 446 passed; 4 failed (compilation errors: 0)
constants ... 471 passed; 4 failed (compilation errors: 0)
constants ... 496 passed; 4 failed (compilation errors: 0)
constants ... 521 passed; 4 failed (compilation errors: 0)
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_3GPP_OPERATOR_CODE" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_3GPP_OPERATOR_NAME" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_3GPP_REGISTRATION_STATE" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_3GPP_SUBSCRIPTION_STATE" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_ACCESS_TECHNOLOGIES" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
constants ... 541 passed; 9 failed (compilation errors: 5)
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_CDMA_CDMA1X_REGISTRATION_STATE" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_CDMA_EVDO_REGISTRATION_STATE" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_CDMA_NID" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_CDMA_SID" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_CURRENT_BANDS" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_SIGNAL_QUALITY" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
compilation command "cc" "-Wno-deprecated-declarations" "-D__USE_MINGW_ANSI_STDIO" "-I/usr/include/libmm-glib" "-I/usr/include/glib-2.0" "-I/usr/lib64/glib-2.0/include" "-pthread" "-I/usr/include/ModemManager" "-DABI_CONSTANT_NAME=MM_SIMPLE_PROPERTY_STATE" "tests/constant.c" "-o" "/tmp/sll/abi.0ZNzM7KYlzWS/constant" failed, exit code: 1
constants ... 559 passed; 16 failed (compilation errors: 12)
constants ... 584 passed; 16 failed (compilation errors: 12)
constants ... 609 passed; 16 failed (compilation errors: 12)
constants ... 634 passed; 16 failed (compilation errors: 12)
constants ... 659 passed; 16 failed (compilation errors: 12)
constants ... 684 passed; 16 failed (compilation errors: 12)
constants ... 709 passed; 16 failed (compilation errors: 12)
thread 'cross_validate_constants_with_c' panicked at 'FAILED: 723 passed; 16 failed (compilation errors: 12)', tests/abi.rs:126:13


failures:
    cross_validate_constants_with_c
    cross_validate_layout_with_c

test result: FAILED. 0 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out

error: test failed, to rerun pass '--test abi'
```

With the current result, I can probably use the sys lib because I don't need any of the failed objects.
I also tried to add API bindings but there were more issues. I guess it's possible to directly use the sys bindings?! Or is it highly recommended to do API bindings?

**Comment by sdroege** [comment]
> So GLib, Gio and GObject should do for "external_libraries" right?

Yes

> I commented out some lines where rust-gir reported a wrong formatting. (at line 1924) I already reported this to the ModemManager maintainer.

How did this happen? This file is supposed to be autogenerated from the C code/headers :) Does MM ship a hand-written .gir file, or is there a bug in gobject-introspection that generates something weird like this?

> Some of the constants cannot be matched because they are present in ModemManager.gir but not actually in the installed libmm-glib.h.

That's another bug in MM. Whatever is in the .gir file should also be in the headers (but not necessarily the other way around).

> The same for MMCall but I assume that rust-gir cannot properly generate the struct because of the problem in ModemManager.gir

As a first step the .gir file should be brought into proper shape and valid, otherwise you'll only run into further problems further down along the line

> Layout mismatch for MMCall Rust: Layout { size: 8, alignment: 8 } C: Layout { size: 48, alignment: 8 }

This is because the struct is wrong in the .gir. The commented out parent in there. That should be fixed :)

> I also tried to add API bindings but there were more issues.

Not unexpected considering the above.

> I guess it's possible to directly use the sys bindings?!

Yes but you'll have to write unsafe Rust then, do memory management yourself, convert types between Rust and C yourself, etc. That's what the API bindings would take care of for you.

> Or is it highly recommended to do API bindings?

For the above reasons, yes :)

----

In summary, first fix the .gir files. The `<!-- <...` lines should not be commented at all, they look correct. The problem seems to be that `MmGdbusCallProxy` is not in the .gir file and thus it can't properly generate that. So probably by adding `MmGdbusCallProxy` this problem would disappear (which I might only require including another header or source file in the gobject-introspection invocation).

**Comment by EPashkin** [comment]
Ignore need be "ModemManager.ElementName", see https://github.com/gtk-rs/sys/blob/master/conf/gir-atk.toml#L14 for consts or https://github.com/gtk-rs/sys/blob/master/conf/gir-gobject.toml#L13 for classes;
it possible reason that you still get warning about ignored in test.

**Comment by EPashkin** [comment]
> I had to update the required package versions to "0.6.0". The default generated was "0.2.0" which didn't work.

if you meant `Cargo.toml` it generated only example, you can change it as you wish, at next regen it can be only minimal update.

**Comment by EPashkin** [comment]
Also maybe you need change "tests\manual.h" to include right header with `MM_SIMPLE_PROPERTY_3GPP_OPERATOR_CODE`,
it also don't be changed on next regen.

**Comment by stfl** [comment]
ok. so 0xFFFFFFFF is actually UINT_MAX which is the same as -1
https://stackoverflow.com/questions/1863153/why-unsigned-int-0xffffffff-is-equal-to-int-1

How should this be treated properly making the tests not fail?

**Comment by EPashkin** [comment]
@stfl "-1" writen it your ModemManager gir file, you can fix it here
```
      <member name="any"
              value="-1"
              c:identifier="MM_MODEM_CAPABILITY_ANY"
              glib:nick="any">
      </member>
```

**Comment by stfl** [comment]
Thanks. the tests for the sys bindings are running through now.

**Comment by EPashkin** [comment]
Seems it time to close issue, as next opened 😉
